### DISCUSSION TITLE
CSS for OSCAL XML

### DISCUSSION THREAD

**Discussion by wendellpiez** [usnistgov/OSCAL#1732] [Ideas]
Pasted below is some very basic CSS providing very basic layout to OSCAL Catalog XML.

How to use:

* At the top of an OSCAL XML catalog place a processing instruction (PI) as follows: `<?xml-stylesheet type="text/css" href="oscal-simple.css"?>`
* Place the CSS below in a file next to the catalog, named `oscal-simple.css`
  * Alternatively, place it where you want and name it what you want, adjusting the `href` path in the PI
* Open the XML file in a browser
* Adjust the CSS for finer effects

```
/* OSCAL namespace for unprefixed elements */
@namespace "http://csrc.nist.gov/ns/oscal/1.0";

/*  everything is block unless it says otherwise */
* { display: block }

/* inline elements overriding the block setting */
a, insert, code, em, i, b, strong, sub, sup, q, img { display: inline }

catalog { padding: 1em }

control { margin: 0.6em 0em; padding: 0.4em; border: thin solid black }

title { font-size: 110%; font-weight: bold }

/* enhancement titles */
control control title { font-size: 105%; font-weight: normal; font-variant: small-caps }

param { margin: 0.6em 0em; padding: 0.6em; border: thin dotted grey; background: whitesmoke }

param:before { font-family: sans-serif; font-size: 80%; content: 'parameter ' attr(id) ': ' }

/* recursive part structure indents */
part { margin-left: 1em }

prop:after { content: attr(value); color: steelblue }

insert:after {  content: '[insert ' attr(type) ' ' attr(id-ref) ']'; font-weight: bold; font-size: 90%; font-style: italic; background-color: whitesmoke }

a { color: blue }

code { font-family: monospace; font-size: 90% }

em, i { font-style: italic }

b, strong { font-weight: bold }

sup { vertical-align: super; font-size: smaller }

sub { vertical-align: sub; font-size: smaller }

q::before {
  content: open-quote;
}
q::after {
  content: close-quote;
}

img:after {  content: '[image ' attr(src)']'; font-weight: bold; font-size: 90%; font-style: italic; color: darkred }

// finis

```

**Comment by aj-stein-nist** [comment]
Quick question: can you also embed inline CSS for "self-contained" OSCAL document (a catalog, SSP) complete with its own styling? I had not thought to even try that, but I assume that goes beyond the boundaries of what the `xml-stylesheet` PI is intended for. I have to go do research now!

**Comment by wendellpiez** [comment]
Yes but no: it needs an XSLT 1.0 shim at minimum, a 'loader'. Consequently it works only from a server (albeit localhost) kind of too bad.

This is essentially what this architecture does: https://dcl.ils.indiana.edu/teibp/

The XSLT in this case would have templates only for showing with the styling in a place where the browser would know what to do (one template), plus small templates for anything that goes beyond CSS as such - for example, writing links into HTML to provide working links. It's a 1KB XSLT kind of XSLT. It could either acquire the CSS from elsewhere or simply rearrange CSS already there in the file some place.

Why not do this all the time? For publishing, the best argument for transforming the entire document instead of just slapping CSS on it is probably the Accessible HTML argument.
